### DISCUSSION TITLE
Dependency Diagrams

### DISCUSSION THREAD

**Discussion by 94sedighi** [fluentassertions/fluentassertions#2072] [General]
As a new contributor, it was a bit difficult to understand the architecture and dependencies. I tried ReShaper and found some cool features that can create project dependency diagrams and type dependency diagrams for each package, like the images below:

- The first image shows the project dependencies.
- The second image shows the type dependencies of the type package

My idea is to create the type dependency diagram of each package and add it to the contribution documents to make it easier for new contributors.

![Project_dependencies](https://user-images.githubusercontent.com/118891407/209097787-cf00484f-175a-43d6-a3e2-d42cc02127b9.png)
![Type_dependencies_of_Types_package](https://user-images.githubusercontent.com/118891407/209097809-f1c4c3e8-17f9-4716-9cc7-5dddf8677ce2.png)

**Comment by dennisdoomen** [comment]
The first diagram can be quite useful, but the second one will change over time. How do you envision us maintaining those?

**Reply by eNeRGy164** [reply]
You could use [LivingDocumentation](https://github.com/eNeRGy164/LivingDocumentation) to generate a PlantUML diagram with [PlantUml.Builder](https://github.com/eNeRGy164/PlantUml.Builder) during the build 😉

**Comment by 94sedighi** [comment]
Of course, the type dependency diagrams of the packages will change over time, but it will be quite useful to have an idea of how the classes are coupled and how they work in general. Will the changes happen every day, every month or every year?

**Comment by dennisdoomen** [comment]
Maybe we can add them to the root read-me? What do you think @jnyrup ?

**Comment by jnyrup** [comment]
If they can be auto-generated by everyone I'm good with it.
I haven't had the need myself, as I usually open up the code in an IDE and browse around, but I know that it can be helpful for many.

**Reply by 94sedighi** [reply]
> If they can be auto-generated by everyone I'm good with it. I haven't had the need myself, as I usually open up the code in an IDE and browse around, but I know that it can be helpful for many.

they can be auto-generated by every one, but you need the tools to auto generate them ...

**Comment by dennisdoomen** [comment]
@94sedighi if you're up to it, would you mind creating a PR to update the read-me with the diagrams? Maybe you can also add a footnote or something like that with instructions how to generate them.

**Reply by 94sedighi** [reply]
> @94sedighi if you're up to it, would you mind creating a PR to update the read-me with the diagrams? Maybe you can also add a footnote or something like that with instructions how to generate them.

I've done it with Resharper and it is not free. I know Visual Studio Enterprise Edition also offers such tools but also these are not free.
@eNeRGy164 i don't know these tools which you have mentioned, but if these are free and you are familiar with them, you can also create a PR.

**Comment by dennisdoomen** [comment]
That's fine. The entire FA team has licenses for Rider and R#.

**Reply by 94sedighi** [reply]
what about the contributors?
They have to update the diagramms, when the change the APIs

**Reply by jnyrup** [reply]
I would prefer if it could be an automated part of the CI pipeline, to ensure that they are always up to date.
If the diagrams are only to be shown on GitHub, they support rendering Mermaid in Markdown
https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

**Reply by 94sedighi** [reply]
> I would prefer if it could be an automated part of the CI pipeline, to ensure that they are always up to date. If the diagrams are only to be shown on GitHub, they support rendering Mermaid in Markdown https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

Where should i add them, the readme is only a text data?
And would it be enough to create a PR where the diagrams are added?
I am not familiar with the CI pipelines aswell xD

**Reply by jnyrup** [reply]
> Where should i add them, the readme is only a text data?

Maybe(?) GitHub Markdown supports rendering Mermaid from a separate file.
The CI could write to that file.

> And would it be enough to create a PR where the diagrams are added?

If the diagrams aren't updated as part CI I imagine they will quick become wrong when code is changed.

> I am not familiar with the CI pipelines aswell xD

The CI process is bootstrapped by https://github.com/fluentassertions/fluentassertions/blob/develop/.github/workflows/build.yml which triggers this [NUKE](https://nuke.build/) program https://github.com/fluentassertions/fluentassertions/blob/develop/Build/Build.cs

**Reply by IT-VBFK** [reply]
Maybe: https://github.com/fralau/mkdocs-mermaid2-plugin/issues/10#issuecomment-681824120

**Reply by dennisdoomen** [reply]
IMO, having some inline mermaid in the read-me is good enough. We just need to make sure we take it into account when we introduce a new class or project. And that doesn't happen _that_ often anymore.

**Reply by jnyrup** [reply]
My thoughts on having mermaid in a separate file is that I guess it would be easier for the tool that generates it to replace an entire file than replacing part of `readme.md`

**Comment by IT-VBFK** [comment]
Now, that we have a possibility to perform git commands from NUKE, this is I think a cool addition.

But we have to agree on how we want this to be implemented.

@94sedighi If you want to try this.. do you need some help? 🤔
